File: src/python/fldrf.py
# Released under Apache 2.0; refer to LICENSE.txt

from math import floor
from math import fmod
from math import frexp

from .fldr import fldr_preprocess_int
from .fldr import fldr_s

# ========================================================================
# Python style implementation

def fldr_preprocess_float_py(a):
    integers = normalize_floats_py(a)
    return fldr_preprocess_int(integers)

def normalize_floats_py(a):
    ratios = [as_integer_ratio_py(x) for x in a]
    Z = max(r[1] for r in ratios)
    return [r[0]*int(Z/r[1]) for r in ratios]

def as_integer_ratio_py(x):
    mantissa, exponent = frexp(x)
    i = 0
    while (mantissa != floor(mantissa)):
        mantissa *= 2.
        exponent -= 1
        i += 1
        if DBL_MAX_WIDTH <= i:
            assert False, 'Failed to converge: %1.5f' % (x,)
    numerator = int(mantissa)
    denominator = 1
    if exponent > 0:
        numerator <<= exponent
    else:
        denominator <<= abs(exponent)
    return (numerator, denominator)

# ========================================================================
# ANSI C99 Implementation (Sketch)

class double_s:
    def __init__(self, mantissa, width, offset, exponent):
        self.mantissa = mantissa
        self.width = width
        self.offset = offset
        self.exponent = exponent

# DBL_MAX from <float.h> typically (2^53-1)*(2^(1023-52)) ~ 2^1024.
DBL_MAX_WIDTH = 1024

def fldr_preprocess_float_c(a):
    n = len(a)
    doubles = normalize_floats_c(a)
    mantissas = [align_mantissa(d) for d in doubles]
    m = binary_sum(mantissas)
    (r, k) = compute_reject_bits(m)

    h = [0] * k
    H = [[-1]*k for _i in range(n+1)]
    for j in range(k):
        d = 0
        for i in range(n):
            idx = j - (k - len(mantissas[i]))
            w = mantissas[i][idx] if 0 <= idx else 0
            h[j] += (w > 0)
            if w > 0:
                H[d][j] = i
                d += 1
        idx = j - (k - len(r))
        w = r[idx] if 0 <= idx else 0
        h[j] += (w > 0)
        if w > 0:
            H[d][j] = n

    return fldr_s(n, m, k, r, h, H)

def normalize_floats_c(a):
    doubles = [as_integer_ratio_c(x) for x in a]
    max_exponent = max(d.exponent for d in doubles)
    for d in doubles:
        offset = max_exponent - d.exponent
        d.offset += offset
        d.width += offset
        d.exponent = max_exponent
    return doubles

def as_integer_ratio_c(x):
    m, exponent = frexp(x)
    i = 0
    while (m != floor(m)):
        m *= 2.
        exponent -= 1
        i += 1
        if DBL_MAX_WIDTH <= i:
            assert False, 'Failed to converge: %1.5f' % (x,)
    mantissa, width = decimal_to_binary(m)
    if 0 < exponent:
        offset = exponent
        exponent = 0
    else:
        offset = 0
        exponent = abs(exponent)
    return double_s(mantissa, width, offset, exponent)

def decimal_to_binary(x):
    assert x == floor(x)
    bits = [0]*DBL_MAX_WIDTH
    width = 0
    while x > 0:
        bits[width] = int(fmod(x, 2))
        x = floor(x/2)
        width += 1
    return bits, width

def align_mantissa(double):
    mantissa = [0] * (double.width + double.offset)
    start = double.width - 1
    for i in range(0, double.width):
        mantissa[start-i] = double.mantissa[i]
    return mantissa

def compute_reject_bits(m):
    if m[0] == 1 and sum(m) == 1:
        k = len(m) - 1
        r = []
    else:
        k = len(m)
        m_pow_2k = [1] + [0] * (k)
        r = binary_sub(m_pow_2k, m)
        assert len(r) <= k
    return r, k

def binary_sum(arrays):
    m = binary_add(arrays[0], arrays[1])
    for i in range(2, len(arrays)):
        m = binary_add(m, arrays[i])
    return m

def binary_add(a, b):
    la = len(a)
    lb = len(b)
    l = max(la, lb)
    c = 0
    x = [0] * (l + 1)
    for i in range(1, l+1):
        ai = a[la - i] if 0 <= la - i else 0
        bi = b[lb - i] if 0 <= lb - i else 0
        x[l+1-i] = ((ai ^ bi) ^ c)
        c = (ai & bi) | (ai & c) | (bi & c)
    if c == 1:
        x[0] = 1
    return binary_trim_inital_zeros(x)

def binary_sub(a, b):
    la = len(a)
    lb = len(b)
    assert lb <= la
    l = max(la, lb)
    c = 0
    x = [0] * (l + 1)
    for i in range(1, l+1):
        ai = a[la - i] if 0 <= la - i else 0
        bi = b[lb - i] if 0 <= lb - i else 0
        x[l+1-i] = ((ai ^ bi) ^ c)
        c = c & (not (ai ^ bi)) | ((not ai) & bi)
    return binary_trim_inital_zeros(x)

def binary_trim_inital_zeros(x):
    j = 0
    while (j < len(x)) and (x[j] == 0):
        j += 1
    return x[j:] if j < len(x) else [0]

File: src/c/fldrf.c
/*
  Name:     fldr.h
  Purpose:  Fast sampling of random integers (floating-point weights).
  Author:   F. A. Saad
  Copyright (C) 2020 Feras A. Saad, All Rights Reserved.

  Released under Apache 2.0; refer to LICENSE.txt
*/

#include <assert.h>
#include <float.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "fldrf.h"
#include "flip.h"

#define DBL_MAX_WIDTH (int)ceil(log2(DBL_MAX))

fldrf_preprocess_t * fldrf_preprocess(double *a, int n) {

    struct double_s *ratios[n];
    for (int i = 0 ; i < n; i++) {
        ratios[i] = as_integer_ratio(a[i]);
    }

    normalize_double_s(ratios, n);

    struct array_s mantissas[n];
    for (int i = 0; i < n; i++) {
        mantissas[i] = align_mantissa(ratios[i]);
    }

    struct array_s m = binary_sum(mantissas, n);
    int k;
    struct array_s r = compute_reject_bits(m, &k);

    int *h = calloc(k, sizeof(*h));
    int *H = calloc((n+1)*k, sizeof(*H));

    int d;
    for(int j = 0; j < k; j++) {
        d = 0;
        for (int i = 0 ; i < n; i++) {
            H[i*k + j] = -1;
            int idx = j - (k - mantissas[i].length);
            bool w = (0 <= idx) ? mantissas[i].items[idx] : 0;
            h[j] += w;
            if (w) {
                H[d*k + j] = i;
                d += 1;
            }
        }
        H[n*k + j] = -1;
        int idx = j - (k - r.length);
        bool w = (0 <= idx) ? r.items[idx] : 0;
        h[j] += w;
        if (w) {
            H[d*k + j] = n;
            d += 1;
        }
    }

    for (int i = 0; i < n; i++) {
        array_s_free(mantissas[i]);
        double_s_free(ratios[i]);
    }

    fldrf_preprocess_t *x = malloc(sizeof(*x));
    x->n = n;
    x->m = m;
    x->k = k;
    x->r = r;
    x->h = h;
    x->H = H;
    return x;
}

void normalize_double_s(struct double_s *d[], int n) {
    int max_exponent = d[0]->exponent;
    for (int i = 1; i < n; i++) {
        if (max_exponent < d[i]->exponent) {
            max_exponent = d[i]->exponent;
        }
    }
    for (int i = 0; i < n; i++) {
        int offset = max_exponent - d[i]->exponent;
        d[i]->offset += offset;
        d[i]->width += offset;
        d[i]->exponent = max_exponent;
    }
}

struct double_s * as_integer_ratio(double x) {
    int exponent;
    double m = frexp(x, &exponent);

    int i = 0;
    while (m != floor(m)) {
        assert(i < DBL_MAX_WIDTH);
        m *= 2.;
        exponent -= 1;
        i++;
    }

    int width;
    struct array_s mantissa = decimal_to_binary(m, &width);

    int offset;
    if (0 < exponent) {
        offset = exponent;
        exponent = 0;
    } else {
        offset = 0;
        exponent = abs(exponent);
    }

    struct double_s *d = malloc(sizeof(*d));
    d->mantissa = mantissa;
    d->width = width;
    d->offset = offset;
    d->exponent = exponent;
    return d;
}

struct array_s decimal_to_binary(double x, int *width) {
    assert(x == floor(x));

    struct array_s bits = array_s_alloc(DBL_MAX_WIDTH);

    int w = 0;
    while (0 < x) {
        bits.items[w++] = fmod(x, 2);
        x = floor(x/2);
    }

    *width = w;
    return bits;
}

struct array_s align_mantissa(struct double_s *d) {
    struct array_s mantissa = array_s_alloc(d->width + d->offset);

    int start = d->width - 1;
    for (int i = 0; i < d->width; i++) {
        mantissa.items[start-i] = d->mantissa.items[i];
    }

    return mantissa;
}

struct array_s compute_reject_bits(struct array_s m, int *k) {

    struct array_s r;
    if ((m.items[0] == 1) && (array_sum(m) == 1)) {
        *k = m.length - 1;
        r.length = 0;
        r.items = NULL;
    } else {
        *k = m.length;
        struct array_s pow_2k = array_s_alloc(*k + 1);
        pow_2k.items[0] = 1;
        r = binary_sub(pow_2k, m);
        array_s_free(pow_2k);
        assert(r.length <= *k);
    }

    return r;
}

struct array_s binary_sum(struct array_s arrays[], int n) {
    if (n == 1) {
        return arrays[0];
    }

    struct array_s m;
    struct array_s m_prev;
    m = binary_add(arrays[0], arrays[1]);
    m_prev = m;
    for (int i = 2; i < n; i ++) {
        m = binary_add(m_prev, arrays[i]);
        array_s_free(m_prev);
        m_prev = m;
    }

    return m;
}

struct array_s binary_add(struct array_s a, struct array_s b) {
    int la = a.length;
    int lb = b.length;
    int l = lb < la ? la : lb;
    int c = 0;
    int length = l+1;
    struct array_s y = array_s_alloc(length);

    for (int i = 1; i < l + 1; i++) {
        int ai = (0 <= la - i) ? a.items[la - i] : 0;
        int bi = (0 <= lb - i) ? b.items[lb - i] : 0;
        y.items[l+1-i] = ((ai ^ bi) ^ c);
        c = (ai & bi) | (ai & c) | (bi & c);
    }
    if (c == 1) {
       y.items[0] = 1;
    }

    remove_leading_zeros(&y);
    return y;
}

struct array_s binary_sub(struct array_s a, struct array_s b){
    int la = a.length;
    int lb = b.length;
    int l = lb < la ? la : lb;
    int c = 0;
    int length = l+1;
    struct array_s y = array_s_alloc(length);

    for (int i = 1; i < l + 1; i++) {
        int ai = (0 <= la - i) ? a.items[la - i] : 0;
        int bi = (0 <= lb - i) ? b.items[lb - i] : 0;
        y.items[l+1-i] = ((ai ^ bi) ^ c);
        c = c & (!(ai ^ bi)) | (!ai & bi);
    }
    if (c == 1) {
       y.items[0] = 1;
    }

    remove_leading_zeros(&y);
    return y;
}

void remove_leading_zeros(struct array_s *a) {
    int j = 0;
    while ((j < a->length) && (a->items[j] == 0)) {
        j++;
    }

    if (j == 0) {
        ;
    } else if (j == a->length) {
        a->length = 0;
        free(a->items);
        a->items = NULL;
    }
    else {
        int length = a->length - j;
        memmove(a->items, a->items + j, length * sizeof(*a->items));
        a->items = realloc(a->items, length * sizeof(*a->items));
        a->length = length;
    }
}

int array_sum(struct array_s a) {
    int x = 0;
    for (int i = 0; i < a.length; i++) {
        x += a.items[i];
    }
    return x;
}

int fldrf_sample(fldrf_preprocess_t *x) {
    if (x->n == 1) {
        return 0;
    }
    int n = x->n;
    int k = x->k;
    int *H = x->H;
    int *h = x->h;
    int c = 0;
    int d = 0;
    while (true) {
        int b = flip();
        d = 2*d + (1 - b);
        if (d < h[c]) {
            int z = H[d*k + c];
            if (z < n) {
                return z;
            } else {
                d = 0;
                c = 0;
            }
        } else {
            d = d - h[c];
            c = c + 1;
        }
    }
}

struct array_s array_s_alloc(int length) {
   struct array_s a;
   a.length = length;
   a.items = calloc(length, sizeof(*a.items));
   return a;
}

void array_s_free(struct array_s x) {
    free(x.items);
}

void double_s_free(struct double_s *d) {
    array_s_free(d->mantissa);
    free(d);
}

void fldrf_free(fldrf_preprocess_t *x) {
    array_s_free(x->m);
    array_s_free(x->r);
    free(x->h);
    free(x->H);
    free(x);
}
